### PR TITLE
refactor(theme): move LayoutProviders to Layout/Provider; composeProviders util

### PR DESCRIPTION
**Pull request by slorber**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

As in other places where we nest components, move `@theme/LayoutProviders` to `@theme/Layout/Provider`

Also use a new `composeProviders` fn to avoid the previous ugly nesting 

Thinking of maybe making ThemeCommon expose a single `ThemeCommonProvider` ? 🤔 

## Test Plan

preview

### CODE REVIEW

File: packages/docusaurus-theme-common/src/utils/reactUtils.tsx [hunk 1]
```diff
@@ -5,7 +5,14 @@
...
 
-import {useCallback, useEffect, useLayoutEffect, useMemo, useRef} from 'react';
+import type {ComponentType, ReactNode} from 'react';
+import React, {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+} from 'react';
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-common/src/utils/reactUtils.tsx:15]:
```suggestion
import React, {
  useCallback,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
  type ComponentType,
  type ReactNode,
} from 'react';
```

**Comment by Josh-Cena** [comment]
> Thinking of maybe making ThemeCommon expose a single `ThemeCommonProvider` ? 🤔

I think it's better this way.

**Comment by slorber** [comment]
> I think it's better this way.

Ok, we can try this

A little bit afraid it removes some flexibility in the end (like the ability to swap one of our provider for something else 🤷‍♂️ ), but not sure to have any good usecase so we'll see if anyone complains

**Comment by Josh-Cena** [comment]
Ah sorry! I didn't make it clear. I meant I prefer the current way.

**Pull request merged**